Selenium tests

ndubetz opened the pull request:
Tests are ready, please review.

jgartee commented on client/app/worlds/worlds.controller.js:
```diff
@@ -16,7 +18,10 @@ angular.module('farseerApp')
 
     $scope.convertToLowerCaseNoSpaces = function(nameToConvert) {
-      return nameToConvert.replace(/ /g, '').toLowerCase();
+      if (nameToConvert !== null && nameToConvert !== undefined) {
```
Is there a built-in function that does this...or could it be extracted into your own method isNullOrUndefined(...), returning a boolean?

robertfmurdock replied:
Technically, you can do a 'truthy' check with it and both null and
undefined return false. Its worth noting that these mean different things
in javascript (one is for explicit omission and the other for indicating a
property that was requested was not defined), so if only one is expected
you can narrow things down.

On Fri, Mar 4, 2016 at 2:01 PM, John Gartee <notifications@github.com>
wrote:

> In client/app/worlds/worlds.controller.js
> <https://github.com/FarseerTeam/Farseer/pull/38#discussion_r55073132>:
>
> > @@ -16,7 +18,10 @@ angular.module('farseerApp')
> >      }
> >
> >      $scope.convertToLowerCaseNoSpaces = function(nameToConvert) {
> > -      return nameToConvert.replace(/ /g, '').toLowerCase();
> > +      if (nameToConvert !== null && nameToConvert !== undefined) {
>
> Is there a built-in function that does this...or could it be extracted
> into your own method isNullOrUndefined(...), returning a boolean?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FarseerTeam/Farseer/pull/38/files#r55073132>.
>

ndubetz replied:
Looking at what this method is trying to do, a 'truthy' check should be sufficient. 
Something like
if(nameToConvert){
  //code
}
will do the trick.

ndubetz replied:
Fixed

jgartee commented on client/app/worlds/worlds.controller.spec.js:
```diff
@@ -132,4 +136,30 @@ describe('Controller: WorldsCtrl', function () {
+      scope.deleteWorldOnUserConfirmation(world);
+      scope.$digest();
+
+      expect(scope.deleteWorldStatus).toBe(1);
```
At some point, we should remove the magic number and replace it with a manifest constant.

ndubetz replied:
Fixed

jgartee commented on e2e/worlds/worlds.spec.js:
```diff
@@ -0,0 +1,265 @@
+
+      var worldList = element(by.id('worldsList'));
+      browser.wait(ec.presenceOf(worldList),1000).then(function(){
+        expect(worldList.getText()).toEqual('Pandora');
```
I think we forgot to validate this was an href.

ndubetz replied:
Fixed

jgartee commented on e2e/worlds/worlds.spec.js:
```diff
@@ -0,0 +1,265 @@
+        browser.wait(ec.presenceOf(worldList),1000).then(function(){
+          var worldList=element.all(by.id('worldsList'));
+
+          expect(worldList.getText()).toEqual(['Pandora','Pillar']);
```
same here...did not validate it is an href

ndubetz replied:
Fixed

jgartee commented on e2e/worlds/worlds.spec.js:
```diff
@@ -0,0 +1,265 @@
+          fail();
+        }).then(done);
+
+        checkIfHtmlElementIsDisplayed(done, '#worldsList input');
```
nice!

jgartee commented:
Is Nate still working on this PR or should I pick it up to bring it to completion?  I'm happy to do so.  @robertfmurdock?

robertfmurdock commented:
He's not at the forge yet, but I'll ask when he gets in

On Wed, Mar 9, 2016 at 8:23 AM, John Gartee <notifications@github.com>
wrote:

> Is Nate still working on this PR or should I pick it up to bring it to
> completion? I'm happy to do so. @robertfmurdock
> <https://github.com/robertfmurdock>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FarseerTeam/Farseer/pull/38#issuecomment-194294582>.
>

ndubetz commented:
So I have made updates per @jgartee 's code review. Are we good to merge this PR, or do we need another set of eyes on it?

vreynolds commented:
just checked out this branch and grunted.... getting the following 3 selenium fails

1) the application opens to the Worlds screen by default and ... when there are existing worlds in the database and... when editing the existing entries changes the hyperlink to a pre-filled text box when the edit icon is clicked
  Message:
    Expected [ '', '' ] to equal [ 'Pandora', 'Pillar' ].
  Stack:
    Error: Failed expectation
        at /Users/Reynolds/git/Farseer/e2e/worlds/worlds.spec.js:173:51
        at Array.forEach (native)
        at process._tickCallback (node.js:401:9)

2) the application opens to the Worlds screen by default and ... when there are existing worlds in the database and... when editing the existing entries changes the text box to a hyperlink containing the new world name when the save icon is clicked
  Message:
    Expected 'Omega' to equal 'PandoraOmega'.
  Stack:
    Error: Failed expectation
        at /Users/Reynolds/git/Farseer/e2e/worlds/worlds.spec.js:193:39
        at Array.forEach (native)
        at process._tickCallback (node.js:401:9)

3) the application opens to the Worlds screen by default and ... when there are existing worlds in the database and... when editing the existing entries resets to the original world name after altering the text box and clicking the undo icon
  Message:
    Expected [ 'Pandora', 'Pillar' ] to equal [ 'Pandora', '' ].
  Stack:
    Error: Failed expectation
        at /Users/Reynolds/git/Farseer/e2e/worlds/worlds.spec.js:215:42
        at Array.forEach (native)
        at process._tickCallback (node.js:401:9)


... thoughts?

ndubetz commented:
I haven't found a definitive reason why the tests would fail. I switched back to the branch and grunted and I didn't get these failures. I did look at the code, and I have a few theories:

1. Each failure seems to be a result of some missing string. None of these missing data values were the result of an execution, but values that were defined in the setup.  These might be intermittent failures.

2. I know Selenium's WebDriver and Firefox have compatibility issues (this is captured in the issues section). If these failures are from running the Selenium tests in Firefox, perhaps they are due to those compatibility issues. 

3. (Unlikely) On a different branch (no_duplicate_worlds), I modified the code in this area so that the text box value would revert to the original world name if an illegal edit was made (duplicate names, for example). I had modified the e2e tests to reflect these changes. I don't know why/how changes from that branch would come here, but I'm still pretty new to git.

robertfmurdock commented:
I just pulled and ran the Farseer:selenium_tests branch and that passes
tests. When I checkout the ndubetz-selenium_tests master branch, I get a
failure. I'm guessing there's some branch confusion in play?

On Fri, Mar 11, 2016 at 10:59 AM, ndubetz <notifications@github.com> wrote:

> I haven't found a definitive reason why the tests would fail. I switched
> back to the branch and grunted and I didn't get these failures. I did look
> at the code, and I have a few theories:
>
>    1.
>
>    Each failure seems to be a result of some missing string. None of
>    these missing data values were the result of an execution, but values that
>    were defined in the setup. These might be intermittent failures.
>    2.
>
>    I know Selenium's WebDriver and Firefox have compatibility issues
>    (this is captured in the issues section). If these failures are from
>    running the Selenium tests in Firefox, perhaps they are due to those
>    compatibility issues.
>    3.
>
>    (Unlikely) On a different branch (no_duplicate_worlds), I modified the
>    code in this area so that the text box value would revert to the original
>    world name if an illegal edit was made (duplicate names, for example). I
>    had modified the e2e tests to reflect these changes. I don't know why/how
>    changes from that branch would come here, but I'm still pretty new to git.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FarseerTeam/Farseer/pull/38#issuecomment-195427655>.
>

robertfmurdock commented:
Possibly not though: I just ran ndubetz-selenium_tests master tests twice
and had two good builds. One with my existing databases, and one after I
dropped all my farseer databases.  Vera, can you confirm that you see the
issue consistently?

On Fri, Mar 11, 2016 at 12:28 PM, Rob Murdock <rmurdock@pillartechnology.com
> wrote:

> I just pulled and ran the Farseer:selenium_tests branch and that passes
> tests. When I checkout the ndubetz-selenium_tests master branch, I get a
> failure. I'm guessing there's some branch confusion in play?
>
> On Fri, Mar 11, 2016 at 10:59 AM, ndubetz <notifications@github.com>
> wrote:
>
>> I haven't found a definitive reason why the tests would fail. I switched
>> back to the branch and grunted and I didn't get these failures. I did look
>> at the code, and I have a few theories:
>>
>>    1.
>>
>>    Each failure seems to be a result of some missing string. None of
>>    these missing data values were the result of an execution, but values that
>>    were defined in the setup. These might be intermittent failures.
>>    2.
>>
>>    I know Selenium's WebDriver and Firefox have compatibility issues
>>    (this is captured in the issues section). If these failures are from
>>    running the Selenium tests in Firefox, perhaps they are due to those
>>    compatibility issues.
>>    3.
>>
>>    (Unlikely) On a different branch (no_duplicate_worlds), I modified
>>    the code in this area so that the text box value would revert to the
>>    original world name if an illegal edit was made (duplicate names, for
>>    example). I had modified the e2e tests to reflect these changes. I don't
>>    know why/how changes from that branch would come here, but I'm still pretty
>>    new to git.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/FarseerTeam/Farseer/pull/38#issuecomment-195427655>.
>>
>
>

robertfmurdock commented:
Also, before we merge, see if you can resolve the issues we're seeing in CI
with those branches:

https://circleci.com/gh/FarseerTeam/Farseer

On Fri, Mar 11, 2016 at 12:34 PM, Rob Murdock <rmurdock@pillartechnology.com
> wrote:

> Possibly not though: I just ran ndubetz-selenium_tests master tests twice
> and had two good builds. One with my existing databases, and one after I
> dropped all my farseer databases.  Vera, can you confirm that you see the
> issue consistently?
>
> On Fri, Mar 11, 2016 at 12:28 PM, Rob Murdock <
> rmurdock@pillartechnology.com> wrote:
>
>> I just pulled and ran the Farseer:selenium_tests branch and that passes
>> tests. When I checkout the ndubetz-selenium_tests master branch, I get a
>> failure. I'm guessing there's some branch confusion in play?
>>
>> On Fri, Mar 11, 2016 at 10:59 AM, ndubetz <notifications@github.com>
>> wrote:
>>
>>> I haven't found a definitive reason why the tests would fail. I switched
>>> back to the branch and grunted and I didn't get these failures. I did look
>>> at the code, and I have a few theories:
>>>
>>>    1.
>>>
>>>    Each failure seems to be a result of some missing string. None of
>>>    these missing data values were the result of an execution, but values that
>>>    were defined in the setup. These might be intermittent failures.
>>>    2.
>>>
>>>    I know Selenium's WebDriver and Firefox have compatibility issues
>>>    (this is captured in the issues section). If these failures are from
>>>    running the Selenium tests in Firefox, perhaps they are due to those
>>>    compatibility issues.
>>>    3.
>>>
>>>    (Unlikely) On a different branch (no_duplicate_worlds), I modified
>>>    the code in this area so that the text box value would revert to the
>>>    original world name if an illegal edit was made (duplicate names, for
>>>    example). I had modified the e2e tests to reflect these changes. I don't
>>>    know why/how changes from that branch would come here, but I'm still pretty
>>>    new to git.
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/FarseerTeam/Farseer/pull/38#issuecomment-195427655>.
>>>
>>
>>
>

vreynolds commented:
Yeah I just ran it once last night, I can definitely try a few more things
when I get back Monday.
On Mar 11, 2016 12:35 PM, "Rob Murdock" <notifications@github.com> wrote:

> Also, before we merge, see if you can resolve the issues we're seeing in CI
> with those branches:
>
> https://circleci.com/gh/FarseerTeam/Farseer
>
> On Fri, Mar 11, 2016 at 12:34 PM, Rob Murdock <
> rmurdock@pillartechnology.com
> > wrote:
>
> > Possibly not though: I just ran ndubetz-selenium_tests master tests twice
> > and had two good builds. One with my existing databases, and one after I
> > dropped all my farseer databases. Vera, can you confirm that you see the
> > issue consistently?
> >
> > On Fri, Mar 11, 2016 at 12:28 PM, Rob Murdock <
> > rmurdock@pillartechnology.com> wrote:
> >
> >> I just pulled and ran the Farseer:selenium_tests branch and that passes
> >> tests. When I checkout the ndubetz-selenium_tests master branch, I get a
> >> failure. I'm guessing there's some branch confusion in play?
> >>
> >> On Fri, Mar 11, 2016 at 10:59 AM, ndubetz <notifications@github.com>
> >> wrote:
> >>
> >>> I haven't found a definitive reason why the tests would fail. I
> switched
> >>> back to the branch and grunted and I didn't get these failures. I did
> look
> >>> at the code, and I have a few theories:
> >>>
> >>> 1.
> >>>
> >>> Each failure seems to be a result of some missing string. None of
> >>> these missing data values were the result of an execution, but values
> that
> >>> were defined in the setup. These might be intermittent failures.
> >>> 2.
> >>>
> >>> I know Selenium's WebDriver and Firefox have compatibility issues
> >>> (this is captured in the issues section). If these failures are from
> >>> running the Selenium tests in Firefox, perhaps they are due to those
> >>> compatibility issues.
> >>> 3.
> >>>
> >>> (Unlikely) On a different branch (no_duplicate_worlds), I modified
> >>> the code in this area so that the text box value would revert to the
> >>> original world name if an illegal edit was made (duplicate names, for
> >>> example). I had modified the e2e tests to reflect these changes. I
> don't
> >>> know why/how changes from that branch would come here, but I'm still
> pretty
> >>> new to git.
> >>>
> >>> —
> >>> Reply to this email directly or view it on GitHub
> >>> <https://github.com/FarseerTeam/Farseer/pull/38#issuecomment-195427655
> >.
> >>>
> >>
> >>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FarseerTeam/Farseer/pull/38#issuecomment-195469714>.
>

ndubetz commented:
@robertfmurdock I was not aware of this CI tool.  I'll see if I can fix the failures it's outputting.
